circus-train-core/src/test/java/com/hotels/bdp/circustrain/core/data/DefaultDataManipulationClientFactoryManagerTest.java

```diff
@@ -0,0 +1,194 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.core.data;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.hadoop.fs.Path;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import com.hotels.bdp.circustrain.api.data.DataManipulationClient;
+import com.hotels.bdp.circustrain.api.data.DataManipulationClientFactory;
+import com.hotels.bdp.circustrain.api.data.DataManipulationClientFactoryManager;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DefaultDataManipulationClientFactoryManagerTest {
+
+  private DataManipulationClientFactory s3s3ClientFactory;
+  private DataManipulationClientFactory s3MapReduceClientFactory;
+  private DataManipulationClientFactory hdfsClientFactory;
+
+  private DataManipulationClientFactoryManager manager;
+  private DataManipulationClientFactory clientFactory;
+  private Path sourceLocation;
+  private Path replicaLocation;
+  private final String s3Path = "s3://<path>";
+  private final String hdfsPath = "hdfs://<path>";
+  private final Map<String, Object> copierOptions = new HashMap<>();
+
+  @Before
+  public void setup() {
+    s3s3ClientFactory = new TestDataManipulationClientFactory();
+    ((TestDataManipulationClientFactory) s3s3ClientFactory).setS3Client();
+
+    s3MapReduceClientFactory = new TestDataManipulationClientFactory();
+    ((TestDataManipulationClientFactory) s3MapReduceClientFactory).setS3MapreduceClient();
+
+    hdfsClientFactory = new TestDataManipulationClientFactory();
+    ((TestDataManipulationClientFactory) hdfsClientFactory).setHdfsClient();
+
+    manager = new DefaultDataManipulationClientFactoryManager(
+        Arrays.asList(s3s3ClientFactory, s3MapReduceClientFactory, hdfsClientFactory));
+    sourceLocation = new Path(s3Path);
+  }
+
+  @Test
+  public void awsClientReturnedForS3S3Copy() {
+    replicaLocation = new Path(s3Path);
+    clientFactory = manager.getClientFactory(sourceLocation, replicaLocation, copierOptions);
+
+    Assert.assertTrue(((TestDataManipulationClientFactory) clientFactory).isS3S3Client());
+  }
+
+  @Test
+  public void awsMapReduceClientReturnedForHdfsS3Copy() {
+    sourceLocation = new Path(hdfsPath);
+    replicaLocation = new Path(s3Path);
+    clientFactory = manager.getClientFactory(sourceLocation, replicaLocation, copierOptions);
+
+    Assert.assertTrue(((TestDataManipulationClientFactory) clientFactory).isS3MapreduceClient());
```

Comment:
Is this checking anything other than the logic you've put in the test class itself? I don't think so?

Reply:
Not really, just checking that it will go through the list of factories, check if the replication is supported and returns the right one. And that it returns the one specified in the copier options (if any).